File: app/src/main/java/com/veganbeauty/admin/features/booking/BookingViewModel.java
package com.veganbeauty.admin.features.booking;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;
import com.veganbeauty.admin.data.local.RootieAdminDatabase;
import com.veganbeauty.admin.data.local.SessionManager;
import com.veganbeauty.admin.data.local.entities.BookingEntity;
import com.veganbeauty.admin.data.remote.FirebaseService;
import com.veganbeauty.admin.data.repository.BookingRepository;
import java.util.ArrayList;
import java.util.List;

public class BookingViewModel extends AndroidViewModel {

    private final BookingRepository repository;
    private final SessionManager sessionManager;

    private final LiveData<List<BookingEntity>> allBookings;
    private final MutableLiveData<String> activeTabStatus = new MutableLiveData<>("PENDING"); // PENDING, UPCOMING, COMPLETED, CANCELLED
    private final MediatorLiveData<List<BookingEntity>> filteredBookings = new MediatorLiveData<>();

    public BookingViewModel(@NonNull Application application) {
        super(application);
        RootieAdminDatabase database = RootieAdminDatabase.getDatabase(application);
        repository = new BookingRepository(database.bookingDao(), new FirebaseService(), application);
        sessionManager = new SessionManager(application);
        allBookings = repository.getAllBookings();

        filteredBookings.addSource(allBookings, bookings -> updateFilteredBookings());
        filteredBookings.addSource(activeTabStatus, status -> updateFilteredBookings());
    }

    public LiveData<List<BookingEntity>> getAllBookings() {
        return allBookings;
    }

    public MutableLiveData<String> getActiveTabStatus() {
        return activeTabStatus;
    }

    public LiveData<List<BookingEntity>> getFilteredBookings() {
        return filteredBookings;
    }

    public void syncFromFirebase() {
        new Thread(() -> {
            try {
                repository.syncFromFirebase();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();
    }

    public void updateBookingStatus(String bookingId, String status, String cancelReason) {
        new Thread(() -> {
            try {
                repository.updateBookingStatus(bookingId, status, cancelReason);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();
    }

    public void updateBookingStatus(String bookingId, String status) {
        updateBookingStatus(bookingId, status, "");
    }

    private void updateFilteredBookings() {
        List<BookingEntity> bookings = allBookings.getValue();
        if (bookings == null) {
            bookings = new ArrayList<>();
        }
        String statusFilter = activeTabStatus.getValue();
        if (statusFilter == null) {
            statusFilter = "PENDING";
        }
        String userRole = sessionManager.getRole();
        if (userRole == null) {
            userRole = "admin";
        }
        String userStoreId = sessionManager.getStoreID();
        if (userStoreId == null) {
            userStoreId = "";
        }
        boolean isStaff = "staff".equalsIgnoreCase(userRole) || "nhân viên".equalsIgnoreCase(userRole);

        // 1. Branch/Store filtering based on role and storeID
        List<BookingEntity> branchFiltered = new ArrayList<>();
        if (isStaff && !userStoreId.isEmpty()) {
            for (BookingEntity be : bookings) {
                boolean storeIdMatches = userStoreId.equals(be.getStoreID());
                boolean storeNameMatches = be.getStoreID() != null && be.getStoreID().isEmpty() &&
                    (("CH001".equals(userStoreId) && be.getStoreName() != null && be.getStoreName().toLowerCase().contains("cơ sở 1")) ||
                     ("CH005".equals(userStoreId) && be.getStoreName() != null && be.getStoreName().toLowerCase().contains("cơ sở 5")));
                if (storeIdMatches || storeNameMatches) {
                    branchFiltered.add(be);
                }
            }
        } else {
            branchFiltered.addAll(bookings);
        }

        // 2. Status filtering based on tab
        List<BookingEntity> statusFiltered = new ArrayList<>();
        for (BookingEntity be : branchFiltered) {
            String status = be.getStatus();
            if (status == null) status = "";
            switch (statusFilter) {
                case "PENDING":
                    if ("Chờ xác nhận".equalsIgnoreCase(status) || "pending".equalsIgnoreCase(status)) {
                        statusFiltered.add(be);
                    }
                    break;
                case "UPCOMING":
                    if ("Sắp diễn ra".equalsIgnoreCase(status) || "confirmed".equalsIgnoreCase(status) || "upcoming".equalsIgnoreCase(status)) {
                        statusFiltered.add(be);
                    }
                    break;
                case "COMPLETED":
                    if ("Đã hoàn thành".equalsIgnoreCase(status) || "completed".equalsIgnoreCase(status)) {
                        statusFiltered.add(be);
                    }
                    break;
                case "CANCELLED":
                    if ("Đã huỷ".equalsIgnoreCase(status) || "Đã hủy".equalsIgnoreCase(status) || "cancelled".equalsIgnoreCase(status)) {
                        statusFiltered.add(be);
                    }
                    break;
                default:
                    statusFiltered.add(be);
                    break;
            }
        }

        filteredBookings.setValue(statusFiltered);
    }
}

File: app/src/main/java/com/veganbeauty/admin/features/booking/BookingViewModel.kt
package com.veganbeauty.admin.features.booking

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.veganbeauty.admin.data.local.RootieAdminDatabase
import com.veganbeauty.admin.data.local.SessionManager
import com.veganbeauty.admin.data.local.entities.BookingEntity
import com.veganbeauty.admin.data.remote.FirebaseService
import com.veganbeauty.admin.data.repository.BookingRepository
import kotlinx.coroutines.launch

class BookingViewModel(application: Application) : AndroidViewModel(application) {

    private val repository: BookingRepository
    private val sessionManager: SessionManager

    val allBookings: LiveData<List<BookingEntity>>
    val activeTabStatus = MutableLiveData<String>("PENDING") // PENDING, UPCOMING, COMPLETED, CANCELLED

    private val _filteredBookings = MediatorLiveData<List<BookingEntity>>()
    val filteredBookings: LiveData<List<BookingEntity>> get() = _filteredBookings

    init {
        val database = RootieAdminDatabase.getDatabase(application)
        repository = BookingRepository(database.bookingDao(), FirebaseService(), application)
        sessionManager = SessionManager(application)
        allBookings = repository.allBookings

        _filteredBookings.addSource(allBookings) { updateFilteredBookings() }
        _filteredBookings.addSource(activeTabStatus) { updateFilteredBookings() }
    }

    fun syncFromFirebase() {
        viewModelScope.launch {
            try {
                repository.syncFromFirebase()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun updateBookingStatus(bookingId: String, status: String, cancelReason: String = "") {
        viewModelScope.launch {
            try {
                repository.updateBookingStatus(bookingId, status, cancelReason)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun updateFilteredBookings() {
        val bookings = allBookings.value ?: emptyList()
        val statusFilter = activeTabStatus.value ?: "PENDING"
        val userRole = sessionManager.getRole() ?: "admin"
        val userStoreId = sessionManager.getStoreID() ?: ""
        val isStaff = userRole.equals("staff", ignoreCase = true) || userRole.equals("nhân viên", ignoreCase = true)

        // 1. Branch/Store filtering based on role and storeID
        var result = if (isStaff && userStoreId.isNotEmpty()) {
            bookings.filter {
                it.storeID == userStoreId ||
                (it.storeID.isEmpty() && (
                    (userStoreId == "CH001" && it.storeName.contains("Cơ sở 1", ignoreCase = true)) ||
                    (userStoreId == "CH005" && it.storeName.contains("Cơ sở 5", ignoreCase = true))
                ))
            }
        } else {
            bookings
        }

        // 2. Status filtering based on tab
        result = when (statusFilter) {
            "PENDING" -> result.filter { 
                it.status.equals("Chờ xác nhận", ignoreCase = true) || 
                it.status.equals("pending", ignoreCase = true) 
            }
            "UPCOMING" -> result.filter { 
                it.status.equals("Sắp diễn ra", ignoreCase = true) || 
                it.status.equals("confirmed", ignoreCase = true) || 
                it.status.equals("upcoming", ignoreCase = true) 
            }
            "COMPLETED" -> result.filter { 
                it.status.equals("Đã hoàn thành", ignoreCase = true) || 
                it.status.equals("completed", ignoreCase = true) 
            }
            "CANCELLED" -> result.filter { 
                it.status.equals("Đã huỷ", ignoreCase = true) || 
                it.status.equals("Đã hủy", ignoreCase = true) || 
                it.status.equals("cancelled", ignoreCase = true) 
            }
            else -> result
        }

        _filteredBookings.value = result
    }
}
